tests/hrex/test_hrex_rbfe.py

```diff
@@ -57,16 +60,30 @@ def test_hrex_rbfe_hif2a(hif2a_single_topology_leg):
     )
     n_windows = 5
 
-    result = estimate_relative_free_energy_bisection_hrex(
-        mol_a,
-        mol_b,
-        core,
-        forcefield,
-        host_config,
-        md_params,
-        lambda_interval=(0.0, 0.15),
-        n_windows=n_windows,
-    )
+    rss_traj = []
+
+    def sample_and_record_rss(*args, **kwargs):
+        traj = sample_with_context(*args, **kwargs)
+        rss_traj.append(Process().memory_info().rss)
+        return traj
+
+    with patch("timemachine.fe.free_energy.sample_with_context", sample_and_record_rss):
+        result = estimate_relative_free_energy_bisection_hrex(
+            mol_a,
+            mol_b,
+            core,
+            forcefield,
+            host_config,
+            md_params,
+            lambda_interval=(0.0, 0.15),
+            n_windows=n_windows,
+        )
+
+    # Check that memory usage is not increasing
+    rss_traj = rss_traj[10:]  # discard initial transients
+    rss_diff_count = np.sum(np.diff(rss_traj) != 0)
+    rss_increase_count = np.sum(np.diff(rss_traj) > 0)
+    assert stats.binom.pmf(rss_increase_count, n=rss_diff_count, p=0.5) >= 0.001
```

Comment:
[nit, non-blocking] I think this detects the scenario we want, and this looks quite robust, very unlikely to lead to false positives, but I was just curious what kinds of corner cases it might miss. Two examples come to mind:

```python
corner_case_traj_false_negative = np.cumsum([[-0.1, +1]]*100)
corner_case_traj_false_positive = np.cumsum([[-0.1, +0.1]]*1000000)
```
![image](https://github.com/proteneer/timemachine/assets/5759036/aabf5db6-fc28-4afe-b668-99c7850ac5bf)
`diff_count = 189, increase_count = 95, binom_pmf=0.05780851636953422`

![image](https://github.com/proteneer/timemachine/assets/5759036/da71c9dc-5ee9-4675-bd8e-2bfbc42ec3cc)
`diff_count = 1999989, increase_count = 999995, binom_pmf=0.0005641909235027816`

--------

```python
from scipy import stats
import numpy as np

def assert_non_increasing(rss_traj, verbose=True):
    rss_traj = rss_traj[10:]  # discard initial transients
    rss_diff_count = np.sum(np.diff(rss_traj) != 0)
    rss_increase_count = np.sum(np.diff(rss_traj) > 0)
    binom_pmf = stats.binom.pmf(rss_increase_count, n=rss_diff_count, p=0.5)
    if verbose:
        print(f"diff_count = {rss_diff_count}, increase_count = {rss_increase_count}, binom_pmf={binom_pmf}")
    assert binom_pmf >= 0.001
```